src/script.c

```diff
@@ -712,6 +712,53 @@ int wally_scriptpubkey_csv_2of2_then_1_from_bytes(
     return WALLY_OK;
 }
 
+int wally_scriptpubkey_csv_2of2_then_1_from_bytes_opt(
+    const unsigned char *bytes, size_t bytes_len, uint32_t csv_blocks,
+    uint32_t flags, unsigned char *bytes_out, size_t len, size_t *written)
+{
+    size_t csv_len = scriptint_get_length(csv_blocks);
+    size_t script_len = 2 * (EC_PUBLIC_KEY_LEN + 1) + 6 + 1 + csv_len; /* 1 for push */
+
+    if (written)
+        *written = 0;
+
+    if (!bytes || bytes_len != 2 * EC_PUBLIC_KEY_LEN ||
+        !csv_blocks || csv_blocks > 0xffff || flags || !bytes_out || !written)
+        return WALLY_EINVAL;
+
+    if (len < script_len) {
+        *written = script_len;
+        return WALLY_OK;
+    }
+
+    /* The script we create is:
+     *     <recovery_pubkey> OP_CHECKSIGVERIFY
+     *     <main_pubkey> OP_CHECKSIG OP_IFDUP OP_NOTIF
+     *         <CSV_BLOCLK> OP_CHECKSEQUENCEVERIFY
+     * OP_ENDIF
+     * Solved by:
+     * 1) The stack containing the main and and recovery signatures.
+     * 2) The stack containing an empty signature and the recovery signature.
+     */
+    *bytes_out++ = EC_PUBLIC_KEY_LEN;
+    memcpy(bytes_out, bytes + EC_PUBLIC_KEY_LEN, EC_PUBLIC_KEY_LEN);
+    bytes_out += EC_PUBLIC_KEY_LEN;
+    *bytes_out++ = OP_CHECKSIGVERIFY;
+    *bytes_out++ = EC_PUBLIC_KEY_LEN;
+    memcpy(bytes_out, bytes, EC_PUBLIC_KEY_LEN);
+    bytes_out += EC_PUBLIC_KEY_LEN;
+    *bytes_out++ = OP_CHECKSIG;
+    *bytes_out++ = OP_IFDUP;
+    *bytes_out++ = OP_NOTIF;
+    *bytes_out++ = csv_len & 0xff;
```

Comment:
If csv blocks is in the interval 1-16, Core rejects the recovery path.

With `testmempoolaccept` I obtained:

    "reject-reason": "non-mandatory-script-verify-flag (Data push larger than necessary)"

This error did not happen if I was spending with the other path.

My guess is that for N in 1-16, we should replace `<1> <n>` with `OP_N`, so for instance:

    >>> wally.scriptpubkey_csv_2of2_then_1_from_bytes_opt(b'\x00'*66, 7, 0).hex()
    '21000000000000000000000000000000000000000000000000000000000000000000ad21000000000000000000000000000000000000000000000000000000000000000000ac73640107b268'
    
should be

    '21000000000000000000000000000000000000000000000000000000000000000000ad21000000000000000000000000000000000000000000000000000000000000000000ac736457b268'

This affects `wally_scriptpubkey_csv_2of2_then_1_from_bytes` and `wally_scriptpubkey_csv_2of3_then_2_from_bytes` too.
Should we fix them so that they produce a script accepted by the mempool too? That may break existing wallets that may not be able to reconstruct past scripts. OTOH, using less than 16 csv blocks seems quite odd and definitely not realistic for meaningful settings, so it may be ok to change it now. Another option is to fail in both cases if csv blocks is 16 or below, which to me is acceptable.

Reply:
Good catch! I think we should fail in this case, its almost certainly going to be an error for a user to set a recovery time this low.